### DISCUSSION TITLE
Efficient way to get size of a graph?

### DISCUSSION THREAD

**Discussion by KonradHoeffner** [pchampin/sophia_rs#116] [Q&A] [answered]
Is counting the iterator `graph.triples().count()` the correct way to get the size of a graph or is there another way?
Right now I only load very small turtle files but I'm worried that when users of my application load let's say a 500 MB Turtle file, that Rust manually going through each item in the iterator may cause a delay, or is that not an issue?

**Comment by pchampin** [comment, marked as answer]
If you are writing generic code, where the only thing you know about `graph` is that it implements the `Graph` trait, then:
* If you need an exact count, yes, `graph.triples().count()` is the way to go.
* If you can make do with an estimation, [`graph.triples().size_hint()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.size_hint) will be cheaper.

The rationale is that the `Graph` trait might be implemented by types that can not cheaply determine the exact number of triples.

If, on the other hand, you know the precise type of `graph`, then you might have other options:
* for all graphs in `sophia::inmem` (`FastGraph`, `LightGraph`), the result of `graph.triples().size_hint()` will always be strictly accurate (min=max).
* `Vec<T>` and `HashSet<T>` implement `Graph` if `T` implements `Triple`; obviously, with those, you can use their `len()` method.
